MennaMohie/assignment3/Solution.cpp

```diff
@@ -66,107 +95,171 @@ class Trie
 		}
 	}
 
-	bool isWord(string word, checkType type = wordCheck)
+	//I need help doing the copy constructor :/
+	/*Trie(const Trie&)
 	{
-		if (word.length() == 0)
-			return false;
 
-		TrieNode *node = root;
+	}
 
-		for (char c : word)
-		{
-			if (!node->hasChild(c))
-				return false;
-			node = node->children[c];
+	Trie& operator =(const Trie&)
+	{
 
-			//if we're doing a wordCheck and we've reached the last letter in the word we need to check its terminates boolean
-			if (type == wordCheck && c == word[word.length() - 1] && node->terminates == false)
-				return false;
-		}
+	}
 
-		return true;
+	//copy constructor is needed first
+	~Trie() 
+	{
+		if(root!=nullptr)
+			remove(root);
+	}*/
+
+	bool isWord(string word)
+	{
+		return check(word, wordCheck);
 	}
 
 	bool isPrefix(string prefix)
 	{
-		return isWord(prefix, prefixCheck);
+		return check(prefix, prefixCheck);
 	}
 };
 
-bool formWords(string word, Trie dictionary, char grid[][gridCols], int rows, int currentRow, int currentCol, vector<string> &words)
+class Finder
 {
-	if (!dictionary.isPrefix(word))
-		return false;
+	Trie dictionary;
+	vector<vector<char>> grid;
+	int rows;
+	int cols;
+	vector<vector<bool>> visited;
 
-	string updatedWord;
-	int updatedRow, updatedCol;
-	for (int x = -1; x <= 1; x++)
+	bool formWords(string &word, int currentRow, int currentCol, vector<string> &words)
 	{
-		for (int y = -1; y <= 1; y++)
+		if (!dictionary.isPrefix(word))
+			return false;
+
+		visited[currentRow][currentCol] = true;
+
+		string updatedWord;
+		int updatedRow, updatedCol;
+		for (int x = -1; x <= 1; x++)
 		{
-			updatedRow = currentRow + x;
-			updatedCol = currentCol + y;
-			if ((y != 0 || x != 0) && updatedRow>=0 && updatedRow<rows && updatedCol>=0 && updatedCol<gridCols)
+			for (int y = -1; y <= 1; y++)
 			{
-				updatedWord = word + grid[currentRow + x][currentCol + y];
-				if (dictionary.isWord(updatedWord))
+				updatedRow = currentRow + x;
+				updatedCol = currentCol + y;
+				if ((y != 0 || x != 0) && updatedRow >= 0 && updatedRow < rows && updatedCol >= 0 && updatedCol < cols)
 				{
-					words.push_back(updatedWord);
+					if (!visited[updatedRow][updatedCol])
+					{
+						updatedWord = word + grid[updatedRow][updatedCol];
+						if (dictionary.isWord(updatedWord))
+						{
+							words.push_back(updatedWord);
+						}
+						formWords(updatedWord, updatedRow, updatedCol, words);
+					}
 				}
-				formWords(updatedWord, dictionary, grid, rows, updatedRow, updatedCol, words);
 			}
 		}
+		return words.size() != 0;
 	}
-	return words.size() != 0;
-}
 
-set<string> formedWords(Trie dictionary, char grid[][3], int rows)
-{
-	set<string> allFormedWords;
-	vector<string> words;
-	string word;
+public:
 
-	for (int r = 0; r < rows; r++)
+	Finder(const Trie &d, const vector<vector<char>> &g, const int r, const int c) :dictionary(d), grid(g), rows(r), cols(c) {}
```

Comment:
r & c parameters are not needed, since you can check the size of the grid directly from g.